Src/Newtonsoft.Json/Utilities/EnumBidirectionalDictionary.cs

```diff
@@ -0,0 +1,102 @@
+﻿using System;
+
+namespace Newtonsoft.Json.Utilities
+{
+    internal class EnumBidirectionalDictionary
+    {
+        private bool _hasCaseSensitiveMembers = false;
+        private readonly BidirectionalDictionary<string, string> _caseInsensitiveDictionary = 
+            new BidirectionalDictionary<string, string>(StringComparer.OrdinalIgnoreCase, StringComparer.OrdinalIgnoreCase);
+
+        private BidirectionalDictionary<string, string> _caseSensitiveDictionary;
+
+        private void InitilizeCaseSensitiveDictionary()
+        {
+            _caseSensitiveDictionary = new BidirectionalDictionary<string, string>(StringComparer.Ordinal, StringComparer.Ordinal);
+        }
+
+        public void Set(string first, string second)
+        {
+            // Add to caseSensitive if second is found, since second is used by TryResolvedEnumName in EnumUtils
+            if (_caseInsensitiveDictionary.TryGetBySecond(second, out string _))
+            {
+                if (!_hasCaseSensitiveMembers)
+                {
+                    _hasCaseSensitiveMembers = true;
+
+                    // Initilize case sensitive dictionary
+                    InitilizeCaseSensitiveDictionary();
+                }
+
+                // Add new element
+                _caseSensitiveDictionary.Set(first, second);
+                
+                // Not needed to add the current element from caseInsentive to caseSensitive since we lookup in _caseInsensitive after _caseSensitive
```

Comment:
Side effect of not moving it is that the first enum will match all other cases.
For example enum with the following members: Abc, abc.
Abc will then match Abc
abc will then match abc
AbC will then match the first

How should I handle this? Should the user be given a exception or should it be first match like now?